src/hb-ot-glyph.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright © 2019-2020  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+#ifndef HB_OT_H_IN
+#error "Include <hb-ot.h> instead."
+#endif
+
+#ifndef HB_OT_GLYPH_H
+#define HB_OT_GLYPH_H
+
+#include "hb.h"
+
+HB_BEGIN_DECLS
+
+typedef void (*hb_ot_glyph_decompose_move_to_func_t) (hb_position_t to_x, hb_position_t to_y, void *user_data);
+typedef void (*hb_ot_glyph_decompose_line_to_func_t) (hb_position_t to_x, hb_position_t to_y, void *user_data);
+typedef void (*hb_ot_glyph_decompose_conic_to_func_t) (hb_position_t control_x, hb_position_t control_y,
+						       hb_position_t to_x, hb_position_t to_y,
+						       void *user_data);
+typedef void (*hb_ot_glyph_decompose_cubic_to_func_t) (hb_position_t control1_x, hb_position_t control1_y,
+						       hb_position_t control2_x, hb_position_t control2_y,
+						       hb_position_t to_x, hb_position_t to_y,
+						       void *user_data);
+
+/**
+ * hb_ot_glyph_decompose_funcs_t:
+ *
+ * Glyph decompose callbacks.
+ *
+ * Since: REPLACEME
+ **/
+typedef struct hb_ot_glyph_decompose_funcs_t
+{
+  hb_ot_glyph_decompose_move_to_func_t move_to;
+  hb_ot_glyph_decompose_line_to_func_t line_to;
+  hb_ot_glyph_decompose_conic_to_func_t conic_to;
+  hb_ot_glyph_decompose_cubic_to_func_t cubic_to;
```

Comment:
Do we need `close` and `end` functions? or how do clients differentiate between closed and open paths (not sure if open paths are allowed at all).

Reply:
I have the `close` for CFF and end for `glyf`, they however not exposed by freetype so guess we can also without them.

Reply:
I prefer having an explicit `close` (and `end` if open contours are possible). Makes client code simpler IMO.

Reply:
Your request is definitely valid, I am trying to see where to insert those :)

We have a start_path and end_path in CFF/CFF2 https://github.com/harfbuzz/harfbuzz/pull/2016/files#diff-7ae1ce6c07f89cffd0c78595d6db8134R359-R360 which perhaps matches `open()` and `end()` or `close()`? am confused more about glyf part, https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L1232-L1242 it wraps each simple glyph with `open()`, `end()`, this https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L1242 however doesn't have begin part weirdly. So I think we can skip those for now if we can draw SVG path without them for now.

fonttools also does a nice thing https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/pens/basePen.py#L187 that has three must override functions and having the rest optional, its `_qCurveToOne` is able to convert quadratic bezier command to cubic, nice thing I guess, perhaps we can provide it (opt-in/default disabled, later?) using some config on the proposed hb_ot_glyph_decompose_funcs_t, wdyt?

Reply:
Okay based on opentype.js's closePath have added open_path and close_path callbacks.

Reply:
And guess turning quadratic bezier to cubic isn't that suitable for us.

Reply:
Have a look please :)